step-automation-packages/step-automation-packages-manager/src/main/java/step/automation/packages/AutomationPackageManager.java

```diff
@@ -259,35 +266,98 @@ public PackageUpdateResult createOrUpdateAutomationPackage(boolean allowUpdate,
+        try {
+            //If not already locked (i.e. was not able to acquire an immediate write lock)
+            if (!alreadyLocked) {
+                log.info("Delaying update of the automation package " + newPackage.getId().toString() + " due to running execution(s) using this package.");
+                getWriteLock(newPackage);
+                log.info("Executions completed, proceeding with the update of the automation package " + newPackage.getId().toString());
+            }
             // delete old package entities
             if (oldPackage != null) {
                 deleteAutomationPackageEntities(oldPackage);
             }
-
             // persist all staged entities
             persistStagedEntities(staging, enricherForIncludedEntities);
-
-            // save automation package metadata
             ObjectId result = automationPackageAccessor.save(newPackage).getId();
-
             logAfterSave(staging, oldPackage, newPackage);
-            return new PackageUpdateResult(oldPackage == null ? PackageUpdateStatus.CREATED : PackageUpdateStatus.UPDATED, result);
-        } catch (Exception ex) {
-            // cleanup created resources
-            try {
-                if (newPackage != null) {
-                    List<Resource> resources = resourceManager.findManyByCriteria(Map.of("customFields." + AutomationPackageEntity.AUTOMATION_PACKAGE_ID, newPackage.getId().toString()));
-                    for (Resource resource : resources) {
-                        resourceManager.deleteResource(resource.getId().toString());
-                    }
+            return result;
+        } finally {
+            if (!alreadyLocked) {
+                releaseWriteLock(newPackage); //only release if lock was acquired in this method
+            }
+            //Clear delayed status
+            newPackage.setStatus(null);
+            automationPackageAccessor.save(newPackage);
+        }
+    }
+
+    private void handleExceptionOnPackageUpdate(AutomationPackage automationPackage) {
+        // cleanup created resources
+        try {
+            if (automationPackage != null) {
+                List<Resource> resources = resourceManager.findManyByCriteria(Map.of("customFields." + AutomationPackageEntity.AUTOMATION_PACKAGE_ID, automationPackage.getId().toString()));
+                for (Resource resource : resources) {
+                    resourceManager.deleteResource(resource.getId().toString());
                 }
-            } catch (Exception e) {
-                log.warn("Cannot cleanup resource", e);
             }
-            throw ex;
+        } catch (Exception e) {
+            log.warn("Cannot cleanup resource", e);
         }
     }
 
+    protected void getWriteLock(AutomationPackage newPackage) {
+        automationPackageLocks.writeLock(newPackage.getId().toHexString());
+    }
+
+    protected boolean getImmediateWriteLock(AutomationPackage newPackage) {
```

Comment:
Consider renaming to `tryGetImmediateWriteLock`, `tryObtainImmediateWriteLock`, `tryImmediateWriteLock` or similar to make the semantics clearer.